Sources/Foundation/FileHandle+Async.swift

```diff
@@ -0,0 +1,207 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if canImport(Darwin)
+import Darwin
+#elseif canImport(Glibc)
+import Glibc
+#elseif canImport(CRT)
+import CRT
+#endif
+
+@available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+final actor IOActor {
+    @available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+    func read(from fd: Int32, into buffer: UnsafeMutableRawBufferPointer) async throws -> Int {
+        while true {
+#if canImport(Darwin)
+            let read = Darwin.read
+#elseif canImport(Glibc)
+            let read = Glibc.read
+#elseif canImport(CRT)
+            let read = CRT._read
+#endif
+            let amount = read(fd, buffer.baseAddress, buffer.count)
+            if amount >= 0 {
+                return amount
+            }
+            let posixErrno = errno
+            if errno != EINTR {
+                // TODO: get the path of the fd to provide a more informative error
+                throw NSError(domain: NSPOSIXErrorDomain, code: Int(posixErrno), userInfo: [:])
+            }
+        }
+    }
+    
+    @available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+    func read(from handle: FileHandle, into buffer: UnsafeMutableRawBufferPointer) async throws -> Int {
+        // this is not incredibly effecient but it is the best we have
+        guard let data = try handle.read(upToCount: buffer.count) else {
+            return 0
+        }
+        data.copyBytes(to: buffer)
+        return data.count
```

Comment:
Is there a specific reason, `handle.readabilityHandler` isn't used here in conjunction with `withUnsafeThrowingContinuation`? Since this is still not part of Foundation on Linux, I wanted to include this code in my package for the moment and just wondered if it was better to use `readabilityHandler` because this function is blocking otherwise.

Reply:
readabilityHandler is less useful than you'd hope for regular files. They're just always readable.

Reply:
But a FileHandle can also be the output of Process, and it may need to wait for new bytes. So the thread will be locked with `read(upToCount:)`. I think `readabilityHandler` will be a better option that we can wrap into _continuation_